File: 1942-the-number-of-the-smallest-unoccupied-chair/1942-the-number-of-the-smallest-unoccupied-chair.java
class Solution {
    public int smallestChair(int[][] times, int targetFriend) {
        int n = times.length;
        Integer[] order = new Integer[n];
        for (int i = 0; i < n; i++) {
            order[i] = i;
        }
        
        Arrays.sort(order, (a, b) -> Integer.compare(times[a][0], times[b][0]));
        
        PriorityQueue<Integer> hpq = new PriorityQueue<>();
        PriorityQueue<int[]> lpq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        
        for (int i = 0; i < n; i++) {
            hpq.add(i);
        }
        
        for (int i : order) {
            int arrival = times[i][0], leave = times[i][1];
            
            while (!lpq.isEmpty() && lpq.peek()[0] <= arrival) {
                hpq.add(lpq.poll()[1]);
            }
            
            int seat = hpq.poll();
            
            if (i == targetFriend) return seat;
            
            lpq.add(new int[]{leave, seat});
        }
        
        return -1;
    }
}

File: 1942-the-number-of-the-smallest-unoccupied-chair/1942-the-number-of-the-smallest-unoccupied-chair.cpp
class Solution {
public:
    int smallestChair(vector<vector<int>>& times, int targetFriend) {
        int n = times.size();
        vector<int> order(n);
        for (int i = 0; i < n; i++) order[i] = i;
        
        auto cmp = [&times](int a, int b) {
            return times[a][0] < times[b][0];
        };
        
        sort(order.begin(), order.end(), cmp);
        
        priority_queue<int, vector<int>, greater<int>> hpq;
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> lpq;
        
        for (int i = 0; i < n; i++) hpq.push(i);
        
        for (int i : order) {
            int arrival = times[i][0], leave = times[i][1];
            
            while (!lpq.empty() && lpq.top().first <= arrival) {
                hpq.push(lpq.top().second);
                lpq.pop();
            }
            
            int seat = hpq.top();
            hpq.pop();
            
            if (i == targetFriend) return seat;
            
            lpq.push({leave, seat});
        }
        return -1;
    }
};